src/main/scala/rise/core/IsClosedForm.scala

```diff
@@ -1,129 +1,45 @@
 package rise.core
 
 import arithexpr.arithmetic.NamedVar
+import rise.core.Traverse._
 import rise.core.traversal.{Continue, Result}
 import rise.core.types._
 
 object IsClosedForm {
-  private case class FreeVariable(v: Any) extends Exception
+  case class Visitor (var boundV: Set[Identifier], var boundT: Set[Kind.Identifier]) extends Traversal[Option] {
+    override implicit def monad = OptionMonad
```

Comment:
This is me trying to be too clever, this implementation is wrong. I will rewrite it and pass the bound variables through the monad.

Reply:
The monad state is just as global as these vars though?  I wonder if we can use a similar style as before here, were the visitor class contains vals and the traversal can be continued locally with a different visitor by re-starting a nested traversal:
```scala
      override def expr(e: Expr): M[Expr] = {
        e match {
          case i: Identifier if !boundV(i) => None
          case Lambda(x, body) =>
            Traverse(body, this.copy(boundV = boundV + x))
          case DepLambda(x, body) =>
            Traverse(body, this.copy(boundT = boundT + x))
          case _ => super.expr(e)
        }
      }
```

Reply:
maybe we can even have some kind of `continue_(e, this.copy(boundV = boundV + x))` implementation which simulates the previous `Continue` behavior?

Reply:
I think I prefer to keep it `Traverse(e, this.copy(...))`, it is more explicit. I was also thinking that we could have a monad bind for just identifier expressions and their corresponding bodies, so that the former can pass information to the latter while at the same time keeping expressions in e.g. application independent, but this is probably overkill.

Reply:
This should now be fixed ;)